src/nifs/protogalaxy/poly/folded_trace.rs

```diff
@@ -0,0 +1,169 @@
+    let rows_count = accumulator.get_witness()[0].len();
+
+    let mut result_matrix_by_challenge = vec![
+        PlonkWitness {
+            W: vec![vec![F::ZERO; rows_count]; columns_count],
+        };
+        X_challenges.len()
+    ];
+
+    itertools::iproduct!(0..columns_count, 0..rows_count)
+        .map(|(col, row)| {
+            iter::once(accumulator.get_witness())
+                .chain(witnesses.iter().map(GetWitness::get_witness))
+                .zip(
+                    lagrange_poly_by_challenge
+                        .iter()
+                        .map(|m| m.iter().copied())
+                        .multi_product(),
+                )
+                .fold(
+                    vec![F::ZERO; X_challenges.len()].into_boxed_slice(),
+                    // every element of this collection - one cell for each `X_challenge`
+                    |mut cells_by_challenge, (witness, multiplier)| {
+                        cells_by_challenge
+                            .iter_mut()
+                            .zip(multiplier.iter())
+                            .for_each(|(res, cell)| {
+                                *res += *cell * witness[col][row];
+                            });
+
+                        cells_by_challenge
+                    },
+                )
+        })
+        .zip(
+            // Here we take an iterator that on each iteration returns [column][row] elements for
+            // each witness for its challenge
+            //
+            // next -> vec![result[0][col][row], result[1][col][row], ... result[challenges_len][col][row]]
+            result_matrix_by_challenge
+                .iter_mut()
+                .map(|matrix| matrix.W.iter_mut().flat_map(|col| col.iter_mut()))
+                .multi_product(),
+        )
+        .par_bridge()
+        .for_each(|(elements, mut results)| {
+            results
+                .iter_mut()
+                .zip(elements.iter())
+                .for_each(|(result, cell)| **result = *cell);
+        });
+
+    result_matrix_by_challenge
+}
+
+fn fold_plonk_challenges<F: PrimeField>(
+    X_challenges: &[F],
+    accumulator: &(impl GetChallenges<F> + Sync),
+    plonk_challenges: &[impl Sync + GetChallenges<F>],
+) -> Vec<Vec<F>> {
+    let log_n = (plonk_challenges.len() + 1).next_power_of_two().ilog2();
+
+    let lagrange_poly_by_challenge = X_challenges
+        .iter()
+        .map(|X| {
+            lagrange::iter_eval_lagrange_polynomials_for_cyclic_group(*X, log_n)
+                .collect::<Box<[_]>>()
+        })
+        .collect::<Box<[_]>>();
+
+    let plonk_challenges_len = accumulator.get_challenges().len();
+
+    iter::once(accumulator.get_challenges())
+        .chain(plonk_challenges.iter().map(GetChallenges::get_challenges))
+        .zip(
+            lagrange_poly_by_challenge
+                .iter()
+                .map(|m| m.iter().copied())
+                .multi_product(),
```

Comment:
Is MultiProduct just hold a collection of iterators? In this case, it is iterator of size "N" (N is number of challenge points):
```
vec![ 
[L_0(X_0), L_1(X_0),...] // first item 
[L_0(X_1), L_1(X_1),...], // second
...
]
```

Reply:
The name of MultiProduct is confusing that I don't know how it related to product

Reply:
This is the name from `itertools`

`MultiProduct` builds a single container with all `next` from a collection of iterators and returns

next[i] -> L_i(X_0), L_i(X_1), ..., L_i(X_n)